worker/src/main/java/gov/cms/ab2d/worker/service/JobProcessingServiceImpl.java

```diff
@@ -128,19 +139,52 @@ public Job processJob(final String jobId) {
         // A mutex lock that all threads for a contract uses while writing into the shared files
         var lock = new ReentrantLock();
 
-        final List<Future<Integer>> futureResourcesHandles = patients.stream()
-                .map(patient -> patient.getPatientId())
-                .filter(patientId -> !isOptOutPatient(patientId))
-                .map(patientId -> patientClaimsProcessor.process(patientId, lock, outputFile, errorFile))
-                .collect(Collectors.toList());
+        int errorCount = 0;
+        JobStatus jobStatus = null;
+
+        int recordsProcessedCount = 0;
+        final List<Future<Integer>> futureResourcesHandles = new ArrayList<>();
+        for (GetPatientsByContractResponse.PatientDTO patient : patients) {
+            ++recordsProcessedCount;
+
+            final String patientId = patient.getPatientId();
+
+            if (isOptOutPatient(patientId)) {
+                // this patient has opted out. skip patient record.
+                continue;
+            }
+
+            futureResourcesHandles.add(patientClaimsProcessor.process(patientId, lock, outputFile, errorFile));
+
+            if (recordsProcessedCount % cancellationCheckFrequency == 0) {
+                errorCount += processHandles(futureResourcesHandles);
+
+                // A Job could run for a long time perhaps hours.
+                // While the job is in progress, the job could be cancelled.
+                // So the worker needs to periodically check the job status to ensure it has not been cancelled.
+
+                jobStatus = jobRepository.findJobStatus(jobUuid);
+                if (jobHasBeenCancelled(jobStatus)) {
+                    log.warn("Job [{}] has been cancelled. Attempting to stop processing the job shortly ... ", jobUuid);
+                    break;
+                }
+
+            }
+        }
 
 
-        int errorCount = processHandles(futureResourcesHandles);
         while (!futureResourcesHandles.isEmpty()) {
             sleep();
             errorCount += processHandles(futureResourcesHandles);
         }
```

Comment:
Even If the job has been cancelled, the flow still goes into this `while` loop that's going to _wait_ for all the Futures to get done. That's the opposite of what we are trying to achieve.

Reply:
Here is what I was thinking.

Assume the contract had 10_000 patients.
The process adds the first `n` patient records (say 10 records = cancellationCheckFrequency) 
to the futures list. 
Every 10 records, it will check the job table to see if it has been cancelled.
If not, it releases whichever future is already done and proceeds with the next 10 records...
check for cancellation again, then proceed with the next 10 records etc.

So, if a cancellation happens, say after the 50th record was processed, there could be about 10 outstanding Futures (give or take a few)  at that point.
So it lets the already issued threads drain, checks their isDone and completes the Futures gracefully.
It would not process any new patient records as it is already in the JobCancellation logic flow.

*TLDR* 
In the `while (!futureResourcesHandles.isEmpty())` loop,  I expect approximately 10 (~= cancellationCheckFrequency, give or take a few) futures to be present - which I figured may be alright to wait on.